component/loki/relabel/relabel.go

```diff
@@ -0,0 +1,235 @@
+		return nil, err
+	}
+
+	c := &Component{
+		opts:         o,
+		metrics:      newMetrics(o.Registerer),
+		cache:        cache,
+		maxCacheSize: args.MaxCacheSize,
+	}
+
+	// Create and immediately export the receiver which remains the same for
+	// the component's lifetime.
+	c.receiver = make(loki.LogsReceiver)
+	o.OnStateChange(Exports{Receiver: c.receiver})
+
+	// Call to Update() to set the relabelling rules once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.receiver:
+			c.metrics.entriesProcessed.Inc()
+			lbls := c.relabel(entry)
+			if len(lbls) == 0 {
+				level.Debug(c.opts.Logger).Log("msg", "dropping entry after relabeling", "labels", entry.Labels.String())
+				continue
+			}
+
+			c.metrics.entriesOutgoing.Inc()
+			entry.Labels = lbls
+			for _, f := range c.fanout {
+				select {
+				case <-ctx.Done():
+					return nil
+				case f <- entry:
+				}
+			}
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	newRCS := flow_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelConfigs)
+	if relabelingChanged(c.rcs, newRCS) {
+		level.Debug(c.opts.Logger).Log("msg", "received new relabel configs, purging cache")
+		c.cache.Purge()
+		c.metrics.cacheSize.Set(0)
+	}
+	if newArgs.MaxCacheSize != c.maxCacheSize {
+		evicted := c.cache.Resize(newArgs.MaxCacheSize)
+		if evicted > 0 {
+			level.Debug(c.opts.Logger).Log("msg", "resizing the cache lead to evicting of items", "len_items_evicted", evicted)
+		}
+	}
+	c.rcs = newRCS
+	c.fanout = newArgs.ForwardTo
+
+	return nil
+}
+
+func relabelingChanged(prev, next []*relabel.Config) bool {
+	if len(prev) != len(next) {
+		return true
+	}
+	for i := range prev {
+		if !reflect.DeepEqual(prev[i], next[i]) {
+			return true
+		}
+	}
+	return false
+}
+
+type cacheItem struct {
+	original  model.LabelSet
+	relabeled model.LabelSet
+}
+
+// TODO(@tpaschalis) It's unfortunate how we have to cast back and forth
+// between model.LabelSet (map) and labels.Labels (slice). Promtail does
+// not have this issue as relabel config rules are only applied to targets.
+// Do we want to use labels.Labels in loki.Entry instead?
+func (c *Component) relabel(e loki.Entry) model.LabelSet {
+	hash := e.Labels.Fingerprint()
+
+	// Let's look in the cache for the hash of the entry's labels.
+	val, found := c.cache.Get(hash)
+	if !found {
+		// We've never seen this hash; initialize it in the cache with a new
+		// cacheItem and return the relabeled result.
+		c.metrics.cacheMisses.Inc()
+		relabeled := c.process(e)
+
+		c.cache.Add(hash, []cacheItem{{e.Labels, relabeled}})
+		c.metrics.cacheSize.Set(float64(c.cache.Len()))
+
+		return relabeled
```

Comment:
Is there a way we can dedupe this with the code below? They're very similar except for the append.

Reply:
Yes, I've gone ahead with a small refactor, hopefully it's easier to follow now.